### PR TITLE
797854 global interest pay bug1

### PR DESCRIPTION
**Pull request by jeanlaroche**
Bug 797854 - Global Register Preference to prompt for interest payment is not being honored.

A new per-account preference is added to let the user decide whether the interest transaction dialog should be
opened automatically before the reconcile. This preference is only enabled for certain types of accounts and
the code that decides that was moved to Account.h as is it now used in two separate place.

### CODE REVIEW

File: libgnucash/engine/Account.cpp [hunk 1]
```diff
@@ -4092,6 +4109,21 @@ xaccAccountGetDescendantPlaceholder (const Account *acc)
...
+gboolean
+xaccAccountGetAutoInterest (const Account *acc)
+{
+    return boolean_from_key (acc, {"auto_interest"});
+}
+
+void
+xaccAccountSetAutoInterest (Account *acc, gboolean val)
+{
+    set_boolean_key (acc, {"auto_interest"}, val);
```
[the remark this answers is not in the document]
Comment (jeanlaroche) [reply libgnucash/engine/Account.cpp:4124]: I'm not following you... Where did I remove that?

[the remark this answers is not in the document]
Comment (jralls) [reply libgnucash/engine/Account.cpp:4124]: You deleted
```
set_boolean_key (acc, {KEY_RECONCILE_INFO, "auto-interest-transfer"}, option);
```
and replaced it with 
```
set_boolean_key (acc, {"auto_interest"}, val);
```

[the remark this answers is not in the document]
Comment (jralls) [reply libgnucash/engine/Account.cpp:4124]: Changing the key from "auto-interest-transfer" to "auto-interest" also breaks compatibility.

[the remark this answers is not in the document]
Comment (jeanlaroche) [reply libgnucash/engine/Account.cpp:4124]: Ah yes! That's a mistake then. I'll fix that.

**Comment by jeanlaroche** [comment]
Should be fixed now.

**Comment by jralls** [comment]
When the edit-account dialog opens the auto-interest-transfer check box is unchecked regardless of whether the KVP is set.

**Comment by jeanlaroche** [comment]
> When the edit-account dialog opens the auto-interest-transfer check box is unchecked regardless of whether the KVP is set.

mmm, I don't see that. How do I reproduce the issue? Are you setting the KVP manually somehow?

**Comment by jralls** [comment]
Nope. Just edit an account, check the box, OK the dialog, edit again: box is unchecked. Close the dialog, reconcile the account and the transfer dialog comes up showing that the KVP is in fact set.

**Comment by jralls** [comment]
Interesting, works fine on Linux. What are you testing on?

**Comment by jeanlaroche** [comment]
> Nope. Just edit an account, check the box, OK the dialog, edit again: box is unchecked. Close the dialog, reconcile the account and the transfer dialog comes up showing that the KVP is in fact set.

Crap, I'm completely unable to repro. It works normally for me. I tried on several accounts, and several books (including some I haven't opened before with this PR)... If I edit the account, tick the check box, close the edit window, reopen it, the check box is checked... The whole things works normally for me. Let me rerun cmake in case something's not in sync.

**Comment by jeanlaroche** [comment]
> Interesting, works fine on Linux. What are you testing on?

Mac os, but I'm testing from the IDE

**Comment by jralls** [comment]
OK, I'll debug into it a bit.

**Comment by jeanlaroche** [comment]
Sorry about that. I just re-ran Cmake in case the glade files were 
misbehaving, but that didn't change anything.

**Comment by jeanlaroche** [comment]
BTW I'd much build on my windows machine, it's a lot more powerful, but I'm put off by the guide that essentially says "don't do it". Is that still the case? Or is building on windows not that hard any longer?

**Comment by jralls** [comment]
It's better than it was, but it's command line all the way. You could always set up a Linux VM.

**Comment by jeanlaroche** [comment]
> It's better than it was, but it's command line all the way. You could always set up a Linux VM.

Ah, no visual C++ or anything like that. Yeah, I'll setup a linux VM and be done with it. Developing on the mac isn't much fun on the mac I'm using, totally underpowered...

**Comment by jralls** [comment]
You can get a lot more mileage out of your mac by building from the command line instead of Xcode, which is quite a hog.

Debugging on MacOS and Linux side-by-side shows that opening the edit account dialog on MacOS fires `gnc_account_type_changed_cb` 4 times and on Linux it fires only once. Each time it fires it calls `set_auto_interest_box`. One of those 4 times it's triggered by `gnc_account_parent_changed_cb` that sets `aw->type` to `ACCOUNT_TYPE_INVALID` and that causes `set_auto_interest_box` to unset the check box. Another call sets `aw->type` back to `ACCOUNT_TYPE_BANK`, but `set_auto_interest_box` can only unset the check box it doesn't get set again before the dialog is shown.

**Comment by jeanlaroche** [comment]
OK I added refresh code to set_auto_interest_box. The drawback is that if you set the flag then change the account type to a type that does not support the flag, the checkbox is unchecked, if you go back to a type that allows it, the flag will not set itself back, because it will just reflect the preference you had before you opened the edit dialog.
This is really not a problem in my view.

**Comment by jralls** [comment]
I agree that it's not a problem, but I think that the following is a bit more straightforward:

```diff
--- a/gnucash/gnome-utils/dialog-account.c
+++ b/gnucash/gnome-utils/dialog-account.c
@@ -157,6 +157,7 @@ void gnc_account_name_insert_text_cb (GtkWidget   *entry,
                                       gint         length,
                                       gint        *position,
                                       gpointer     data);
+static void set_auto_interest_box(AccountWindow *aw);

 /** Implementation *******************************************************/

@@ -278,9 +279,7 @@ gnc_account_to_ui(AccountWindow *aw)
     gtk_toggle_button_set_active (GTK_TOGGLE_BUTTON (aw->hidden_button),
                                   flag);

-    flag = xaccAccountGetAutoInterest (account);
-    gtk_toggle_button_set_active (GTK_TOGGLE_BUTTON (aw->auto_interest_button),
-                                  flag);
+    set_auto_interest_box (aw);
     LEAVE(" ");
 }

@@ -1098,12 +1097,13 @@ static void
 set_auto_interest_box(AccountWindow *aw)
 {
     Account* account = aw_get_account (aw);
-    gboolean flag = account_type_has_auto_interest_xfer (aw->type);
-    gtk_widget_set_sensitive (GTK_WIDGET (aw->auto_interest_button), flag);
-    if (!flag)
-        gtk_toggle_button_set_active (GTK_TOGGLE_BUTTON (aw->auto_interest_button), FALSE);
-    gtk_widget_set_sensitive (GTK_WIDGET (aw->auto_interest_button_label), flag);
-
+    gboolean type_ok = account_type_has_auto_interest_xfer (aw->type);
+    gboolean pref_set = xaccAccountGetAutoInterest (account);
+    gtk_toggle_button_set_active (GTK_TOGGLE_BUTTON (aw->auto_interest_button),
+                                  type_ok && pref_set);
+    gtk_widget_set_sensitive (GTK_WIDGET (aw->auto_interest_button), type_ok);
+    gtk_widget_set_sensitive (GTK_WIDGET (aw->auto_interest_button_label),
+                              type_ok);
 }

 static void
```

**Comment by jeanlaroche** [comment]
> I agree that it's not a problem, but I think that the following is a bit more straightforward:
I agree will make the changes.

**Comment by jeanlaroche** [comment]
I adopted your suggestions. Annoyingly I tried to use your diff as a git patch but wasn't successful in applying it. Not sure what git was complaining about. I probably don't know of a very easy way to do that... :)
J.

**Comment by jralls** [comment]
> Annoyingly I tried to use your diff as a git patch but wasn't successful in applying it. Not sure what git was complaining about.

That's because my diff was against the PR before your refresh commit. I'd think you'd have gotten a fairly simple conflict to resolve.

**Comment by jeanlaroche** [comment]
> > Annoyingly I tried to use your diff as a git patch but wasn't successful in applying it. Not sure what git was complaining about.
> 
> That's because my diff was against the PR before your refresh commit. I'd think you'd have gotten a fairly simple conflict to resolve.

Possibly, but git wouldn't even let me apply it. I was only able to apply part of it. Git would say "corrupt patch at line $" (usually the last line). But no worries.

**Pull request merged**